### ISSUE TITLE
Allow explicit public on members to emit public in .NET metadata even in internal types

### ISSUE THREAD

**Issue by dsyme** [fsharp/fslang-suggestions#1057]
I propose we modify F# default .NET codegen to make an *explicit* use of `public` on a member at the point of implementation cause the corresponding .NET member to always get the "public" attribute even if in an internal type.  This would apply even if the member is completely hidden by a signature (which, in the absence of an explicit public, would cause the emit of internal)

For example

```fsharp
type internal C() = 
     member public x.P
```

The same rule would apply to types in internal/private modules.

```fsharp
module internal M =
    type public C() =    // nested type - now gets .NET public
         member public x.P    // member - now gets .NET public
```

This may plausibly be considered a breaking change, since it changes the behaviour of .NET reflection w.r.t. the type we will need to discuss.  It is possible that a `CLIPublic` attribute will need to be used instead (naming taken from `CLIEvent`)

```fsharp
type internal C() = 
     [<CLIPublic>]
     member x.P
```

For background and discussion see here:

* https://github.com/dotnet/fsharp/issues/11953#issuecomment-896878201

There is no existing way of approaching this problem (reflection over "public" members of private types) in F#.  You have to use reflection over private members.

## Pros and Cons 

The advantages of making this adjustment to F# are ability to emit code that follows .NET norms

The disadvantages of making this adjustment to F# are (see rant in https://github.com/dotnet/fsharp/issues/11953#issuecomment-896878201)

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): M (because of compat concerns)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [ ] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [ ] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by charlesroddie** [comment]
> the very idea of .NET allowing so-called "public" methods in "private" types is a ridiculous notion and should never have been allowed in the design of .NET (Hint: They Are Not Public, They Can't Be Accessed).

With apologies for jumping in out of my depth, I don't get this.

```fsharp
type private A() =
    member _.X = ()

type private B() =
    member private _.Y = ()

let x = A().X // allowed because X is public
let y = B().Y // not allowed because Y is private
```

If X isn't a public method, what is it?

**Comment by dsyme** [comment]
> If X isn't a public method, what is it?

I updated your code to use "internal" since the answer is then clearer.

Here `type internal` means "private to the assembly" .  So `A.X` has the same logical accessibility as `A` - that is, it is not public, but rather internal, since it can only be accessed by things internal to the assembly. 

Thus:
```fsharp
type internal A() =
    member _.X = ()

type internal B() =
    member internal _.Y = ()

type internal C() =
    member private _.Z = ()

let x1 = A().X // allowed because X it's logical visibility is **internal**
let x2 = B().Y // allowed because Y is internal
let x3 = C().Z // not allowed because Z is private
```

**Comment by vzarytovskii** [comment]
Can CLIPublic be used on members with private/internal modifiers?

**Comment by dsyme** [comment]
> Shall this be disallowed?

I don't think so.  You can use it on any member or nested type (including F# types in modules)

**Comment by stmax82** [comment]
> In the meantime I'd encourage Dapper to consider some kind of change to stop the requirement to have things-labelled-public-in-IL-that-arent-actually-public-because-they're-in-a-private-type.

I guess for a library that uses reflection that'd involve something like

```fsharp
let t = someobj.GetType()
if t.IsNestedPrivate then
    t.GetProperties(Public ||| NonPublic)
else
    t.GetProperties(Public)
```

Though I don't think that's a safe solution, since .NET still allows private types to have public properties.. what if a private type has both public and private properties? The author's intention might be that one should only access its publics, but stay away from its privates.. which would turn this into something like:

```fsharp
let t = someobj.GetType()
if t.HasPublicProperties then
    t.GetProperties(Public)
elif t.IsNestedPrivate then
    t.GetProperties(NonPublic)
else []
```

... seems quite complicated and comes with several assumptions to do the right thing for both .NET and F# (also it's probably still wrong).

EDIT: just verified, it really is wrong because IsNestedPrivate is false for both `type private Foo = { ... }` as well as for `type Foo = { ... }` contained in a `module private ... =`... but whatever... proves the point this is complicated to get right for both F# and .NET.

**I think a library (like Dapper) that receives "some object" shouldn't have to mind the type's accessibility, if all it needs are the type's properties.** The easiest and safest thing for a library to do would be to stick to only touching an object's publics, because one way or another they have been marked as public by their author.

> Because of compat concerns it's likely that we'd need a CLIPublic attribute

Just my opinion but I'd prefer the public over attributes because it reads better and both `type Foo = private { ... }` as well as `type Foo = public { ... }` are already supported by the parser (even if the later is silently ignored right now).

As for compat concerns I wonder if any F# code actually uses public on private types right now.. I doubt that there's a lot such code, if any at all, since it currently has no effect / is ignored.

Whichever you pick (public or attribute), I think `type private Foo = public { ... }` should either do what the code says (make the members public), or emit a compiler error / warning. Right now the public is silently ignored (which took us some time to figure out why the code didn't work as intended), while `type Foo = private { ... }` does have an effect already.

**Comment by vzarytovskii** [comment]
> > Shall this be disallowed?
> 
> I don't think so. You can use it on any member or nested type (including F# types in modules)

Gotcha, so attribute shall take precedence over any member modifiers?

**Comment by dsyme** [comment]
> As for compat concerns I wonder if any F# code actually uses public on private types right now.. I doubt that there's a lot such code, if any at all, since it currently has no effect / is ignored.

For compat, there is "some" code, which is too much.  And undoubtedly some will use reflection.

**Comment by GapanovichIgor** [comment]
It looks like SRTP is also broken by this issue. In the example below in all cases it is the same error:
"The member or object constructor 'get_Foo' is not public. Private members may only be accessed from within the declaring type. Protected members may only be accessed from an extending type and cannot be accessed from inner lambda expressions."
```fsharp
module InlineFunction =
    let inline getFoo x = (^a: (member Foo: int) x)

module UseSite1 =
    type internal MyType() =
        member this.Foo = 1

        member this.Test() = InlineFunction.getFoo this // Error

    let test () =
        let value = MyType()

        InlineFunction.getFoo value // Error

module internal UseSite2 =
    type MyType() =
        member this.Foo = 1

        member this.Test() = InlineFunction.getFoo this // Error

    let test () =
        let value = MyType()

        InlineFunction.getFoo value // Error
```
Tested in a library project on .NET 6.
The cure is the same: make the type public all the way up.

So I think this issue needs a proper resolution and not just a "fix it" attribute.

**Comment by dsyme** [comment]
@GapanovichIgor That is covered by #230
